server/lib/chat/chat-service.ts

```diff
@@ -217,27 +206,40 @@ export default class ChatService {
         "Can't moderate yourself",
       )
     }
+
+    const { channelPermissions: moderatorChannelPermissions } = moderatorUserChannel
+    const { channelPermissions: targetChannelPermissions } = targetUserChannel
+
+    // TODO(2Pac): Really need tests for these.
+
+    if (targetChannelPermissions.owner && !moderatorPermissions?.moderateChatChannels) {
+      throw new ChatServiceError(
+        ChatServiceErrorCode.ModeratorAccess,
+        'Only global moderators can moderate channel owners',
+      )
+    }
     if (
+      (targetChannelPermissions.editPermissions ||
+        targetChannelPermissions.ban ||
+        targetChannelPermissions.kick) &&
       !moderatorPermissions?.moderateChatChannels &&
-      !moderatorUserChannel.channelPermissions.owner
+      !moderatorChannelPermissions.owner
     ) {
-      if (!moderatorUserChannel.channelPermissions[moderationAction]) {
-        throw new ChatServiceError(
-          ChatServiceErrorCode.ModeratorAccess,
-          'Not enough permissions to moderate the user',
-        )
-      }
-      if (
-        targetUserChannel.channelPermissions.owner ||
-        targetUserChannel.channelPermissions.editPermissions ||
-        targetUserChannel.channelPermissions.ban ||
-        targetUserChannel.channelPermissions.kick
-      ) {
-        throw new ChatServiceError(
-          ChatServiceErrorCode.ModeratorAccess,
-          "Can't moderate users that have moderator access",
-        )
-      }
+      throw new ChatServiceError(
+        ChatServiceErrorCode.ModeratorAccess,
```

Comment:
so here in this function you are returning the same error code for *3* different errors. We *never ever ever* want to do that, error codes are here to help us differentiate what happened. If these are the same error, check them in the same statement and return the same message as well. If these are different errors, give them different codes.